### DISCUSSION TITLE
Support joining with union expressions

### DISCUSSION THREAD

**Discussion by belovaf** [diesel-rs/diesel#4074] [Ideas]
Currently `InternalJoinDsl` is not implemented for `CombinationClause` and I can't write this:

schema:
```
diesel::table! {
    tasks (id) {
        id -> Int8,
        title -> Text,
    }
}

diesel::table! {
    task_user_attrs (user_id, task_id) {
        #[max_length = 255]
        user_id -> Varchar,
        task_id -> Int8,
        is_viewed -> Bool,
    }
}

diesel::joinable!(task_user_attrs -> tasks (task_id));

diesel::allow_tables_to_appear_in_same_query!(
    task_user_attrs,
    tasks,
);
```

Model:
```
#[derive(Debug, Queryable, Selectable, Serialize)]
#[diesel(table_name = crate::schema::task_user_attrs)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct TaskAttrs {
    is_viewed: bool,
}

#[derive(Debug, Queryable, Selectable, Serialize)]
#[diesel(table_name = crate::schema::tasks)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct Task<'a> {
    pub id: i64,
    title: Cow<'a, str>
```

Usage:
```
let mut con = ...

let combination = tasks.union_all(tasks);

let query = combination
        .left_join(task_user_attrs) // error here!
        .select((Task::as_select(), Option::<TaskAttrs>::as_select()));

let results = query.load::<(Task<'_>, Option<TaskAttrs>)>(&mut con).await.unwrap();
```

Is there any workaround for this?

**Comment by weiznich** [comment]
This is currently not supported by the built-in DSL. It's just something nobody spend time on implementing yet. I think we might be open to accept a well written PR including documentation + tests here.

Otherwise you can always build your own DSL extension or fall back to `diesel::sql_query` for such unsupported cases.

**Reply by dessalines** [reply]
Is this also fixed by #4245 ? 

I'm trying to return multiple data types using `union_all` by:

- Selecting id and a `published` date column from different tables.
- Combine them with `union_all`
- Sort by published desc
- `left_join` to other tables to get other necessary data.

My other option is just to select from each table individually, then combine them in rust code.

**Reply by weiznich** [reply]
@dessalines  No #4245 only implements support for offset and limit. This does not give you support for order or joins, these would require additional traits.

**Reply by dessalines** [reply]
Gotcha, thx. I think I'm going to go about this in a different way then.
